matching-notification-api/src/main/java/org/phenotips/matchingnotification/storage/internal/DefaultMatchStorageManager.java

```diff
@@ -138,125 +103,166 @@
+        Map<PatientSimilarityView, PatientMatch> matchMapping =
+                this.convertSimilarityViewsToPatientMatches(
+                        similarityViews, referenceServerId, matchedServerId, filter);
+
         Session session = this.beginTransaction();
         boolean transactionCompleted = false;
 
         try {
-            Set<String> refPatients = new HashSet<>();
-            refPatients.add(patientId);
+            // load existing matches between this patient and other patients on the matchedServerId
+            List<PatientMatch> existingMatches =
+                    this.loadMatchesForPatientAndServer(patientId, referenceServerId, matchedServerId);
+
+            // to speed up searches group matches by the ID of the other (not patientId) patient in a match
+            Map<String, PatientMatch> matchesByMatchedPatient = new HashMap<>();
+            for (PatientMatch match : existingMatches) {
+                if (match.getReferencePatientId().equals(patientId)) {
+                    matchesByMatchedPatient.put(match.getMatchedPatientId(), match);
+                } else {
+                    matchesByMatchedPatient.put(match.getReferencePatientId(), match);
+                }
+            }
+
+            // there are 4 cases:
+            //
+            // 1) existing match is equivalent (has same match data) to one of the new matches
+            //     -> do nothing with the DB
+            //     -> put existing match to matchMapping so that it has the right ID
+            //
+            // 2) existing match is similar to one of the new matches, but not the same
+            //     -> copy metadata to new match (found timestamp, notes, history, etc.)
+            //     -> copy old match to history
+            //     -> remove old match
+            //     -> save new match
+            //
+            // 3) existing match has no equivalent among new matches
+            //     -> copy old match to history
+            //     -> remove old match
+            //
+            // 4) new match has no equivalent among existing matches
+            //     -> save new match
 
             List<PatientMatch> matchesToSave = new LinkedList<>();
-            for (PatientSimilarityView similarityView : similarityViews) {
-                String refPatientID = similarityView.getReference().getId();
-                if (!patientId.equals(refPatientID)) {
-                    refPatients.add(refPatientID);
-                    if (isIncoming) {
-                        this.logger.error(
-                            "A list of incoming matches for remote patient {} also constains matches for patient {}",
-                            patientId, refPatientID);
+
+            for (Map.Entry<PatientSimilarityView, PatientMatch> entry : matchMapping.entrySet()) {
+                PatientMatch match = entry.getValue();
+                if (matchesByMatchedPatient.containsKey(match.getMatchedPatientId())) {
```

Comment:
NPE when `match = null` from [this](https://github.com/phenotips/patient-network/pull/261/files#diff-01f699dad56819dad6cc877f31af51cbR140) line